src/Swarm/Language/LSP/Hover.hs

```diff
@@ -0,0 +1,167 @@
+  -- | default
+  Syntax ->
+  -- | position
+  Int ->
+  -- | next element to inspect
+  Syntax ->
+  -- | alternate element to inspect
+  Syntax ->
+  Syntax
+descend2 s0 pos s1@(Syntax l1 _) s2 =
+  if withinBound pos l1
+    then narrowToPosition s1 pos
+    else descend s0 pos s2
+
+-- | Find the most specific term for a given
+-- position within the code.
+--
+-- In the case statement, all of the constructors
+-- with nested Syntax files are given explicit
+-- cases, whereas a catchall for non-recursive
+-- constructors is placed at the end.  Note that
+-- this is catch-all is "dangerous" if we ever
+-- add more recursive terms to the language; the
+-- compiler won't warn us about missing cases.
+narrowToPosition ::
+  -- | parent term
+  Syntax ->
+  -- | absolute offset within the file
+  Int ->
+  Syntax
+narrowToPosition s0@(Syntax _ t) pos = case t of
+  SLam _ _ s -> descend s0 pos s
+  SApp s1 s2 -> descend2 s0 pos s1 s2
+  SLet _ _ _ s1 s2 -> descend2 s0 pos s1 s2
+  SPair s1 s2 -> descend2 s0 pos s1 s2
+  SDef _ _ _ s -> descend s0 pos s
+  SBind _ s1 s2 -> descend2 s0 pos s1 s2
+  SDelay _ s -> descend s0 pos s
+  _ -> s0
+
+-- | Markdown line that captures tree depth
+data DocLine a = DocLine
+  { depth :: Int
+  , txt :: a
+  }
+
+renderDoc :: DocLine Text -> Text
+renderDoc (DocLine d t)
+  | d == 0 = t
+  | d == 1 = "* " <> t
+  | otherwise = indent (d - 1) <> "* " <> t
+ where
+  indent x = T.replicate (4 * x) " "
+
+instance Functor DocLine where
+  fmap f (DocLine d x) = DocLine d $ f x
+
+pureDoc :: a -> DocLine a
+pureDoc = DocLine 0
+
+treeToMarkdown :: Int -> Tree (DocLine Text) -> Text
+treeToMarkdown d (Node (DocLine _n t) children) =
+  T.unlines $ renderDoc (DocLine d t) : map (treeToMarkdown $ d + 1) children
+
+explain :: Term -> Tree (DocLine Text)
+explain = \case
+  TUnit -> pure $ pureDoc "The unit value."
+  TConst c -> pure $ pureDoc $ briefDoc $ constDoc $ constInfo c
+  TDir {} -> pure $ pureDoc "A direction literal."
+  TInt {} -> pure $ pureDoc "An integer literal."
+  TAntiInt {} -> pure $ pureDoc "An antiquoted Haskell variable name of type Integer."
+  TText {} -> pure $ pureDoc "A text literal."
+  TAntiText {} -> pure $ pureDoc "An antiquoted Haskell variable name of type Text."
+  TBool {} -> pure $ pureDoc "A Boolean literal."
+  TRobot {} -> pure $ pureDoc "A robot reference.  These never show up in surface syntax, but are here so we can factor pretty-printing for Values through pretty-printing for Terms."
+  TRef {} -> pure $ pureDoc "A memory reference.  These likewise never show up in surface syntax but are here to facilitate pretty-printing."
+  TRequireDevice {} -> pure $ pureDoc "Require a specific device to be installed."
+  TRequire {} -> pure $ pureDoc "Require a certain number of an entity."
+  TVar x -> pure $ pureDoc $ "var: " <> U.bquote x
+  SLam {} -> pure $ pureDoc "A lambda expression, with or without a type annotation on the binder."
+  SApp (Syntax _ t1) (Syntax _ t2) ->
```

Comment:
I'm now rendering function application as a tree:

![f](https://user-images.githubusercontent.com/261693/211168875-d0b04440-8ae5-4484-b501-04592bd02fbe.png)

Reply:
This seems like overkill to me.  It's a lot of information and it's not immediately clear how the nested bullet points correspond to the syntax.  What's the intended audience for these hover tips?

Reply:
My original motivation was that when hovering over operators, all it would say about them was "function application".  So I wanted to see if I could get a bit more detail out of it, e.g. for `==`, now it at least mentions equality:
![Screenshot from 2023-01-07 12-36-55](https://user-images.githubusercontent.com/261693/211169539-e468387f-5035-473a-ba74-e935625d5528.png)

However, for functions of multiple arguments it does get messy.  So this is still experimental.  I would kind of like to commit the progress on this tree rendering and see if I can wrangle some more useful output from it.